### ISSUE TITLE
Launcher: Improve GUI

### ISSUE THREAD

**Issue by Manuel-K** [ja2-stracciatella/ja2-stracciatella#607]
Continuation of #548 

[ ] When switching from predefined to custom resolution, the custom resolution should be updated to the selected predefined resolution if it was 0 x 0.
[ ] The game does not support all possible custom resolutions. We should warn the user if the selected resolution is too low.
[ ] Make it prettier. How? We need mockups/suggestions.

**Comment by JAsmine-ja2** [comment]
Some Mockups back from the Bitbucket days:
https://bitbucket.org/gennady/ja2-stracciatella/wiki/Drawing%20Board

**Comment by Manuel-K** [comment]
@selaux I don't think it was a good idea to remove the fl file when you commited the generated cc and h files.

**Comment by Manuel-K** [comment]
Current layout:
![launcher_0](https://user-images.githubusercontent.com/775439/32417541-776cc2ac-c25b-11e7-9d08-ddb0c40e211a.png)

Minor improvement:
![launcher_1](https://user-images.githubusercontent.com/775439/32417543-799c0b50-c25b-11e7-8f1e-e620d9fb91e2.png)

**Comment by Manuel-K** [comment]
Any opinions?

**Comment by JAsmine-ja2** [comment]
Does the user have to select the game version or does the launcher detect the game version?

Should the game version include the version no besides the language?
e.g. 1.12 "Gold" English
or 1.05 Deutsch

**Comment by Manuel-K** [comment]
The user has to select the game version manually. For languages other than Russian which has the options RUSSIAN and RUSSIAN_GOLD, it's basically a language selector. 

At the moment the options are all caps i.e. they match the possible settings in the config (`ja2.json`) file. I'd prefer normal spelling.

**Comment by JAsmine-ja2** [comment]
Does/Could the launcher use our custom logo as icon?

**Comment by selaux** [comment]
I would like to add vanilla data directory and game version auto detection some time in the future, but for now I think that would be too complicated. For now it should reflect what we can actually set. 

Branding imho is a must (at least use the logo). Also currently resizing does not work properly.

@Manuel-K Minor improvements look good.

Also we have some mocks in a very old ticket: #19 or here https://bitbucket.org/gennady/ja2-stracciatella/wiki/Drawing%20Board.

**Comment by lynxlynxlynx** [comment]
Can we adjust control padding? Looks too small now for the input boxes.

It'd be also great if we had a dropdown with the bundled mods, but that's trickier to do right. Most require to be used with every run, so we'd need to remember. And then you cn have saves with different mods, but by then, it's already too late.

**Comment by Manuel-K** [comment]
Resizing is not implemented at all. I'm not sure how to handle it in FLTK – I'm used to the way Qt does it with the various layouts that scale everything automatically.

**Comment by lynxlynxlynx** [comment]
we can just disable resizing instead of bothering with it.

**Comment by Manuel-K** [comment]
Resizing should already be disabled.

> Can we adjust control padding? Looks too small now for the input boxes.

Can you clarify what you mean? Do you mean the space between the border of the box and the first character. That does seem to be a bit small, but I haven't found a way to change it.

**Comment by Manuel-K** [comment]
More minor improvements:
![launcher_2](https://user-images.githubusercontent.com/775439/32453170-db17bd96-c31b-11e7-9122-96d38a71b48b.png)



I've made the resolution boxes larger and made the “x” centered. 640x480 is now the default value of the input boxes. Also made the “Resolution“ text bold.

**Comment by Manuel-K** [comment]
Setting the icon seems to be rather convoluted in FLTK. 

I've managed to set the main window icon by converting a 32x32 pixel png file of the logo to a char array. This array can be used in `Launcher.cc` to set the icon. Obviously now the launcher needs support for png files.

**Comment by Manuel-K** [comment]
With icon:
![launcher_3](https://user-images.githubusercontent.com/775439/32455559-e816fc8a-c322-11e7-9c47-e095ce531290.png)


Still missing: Windows exe icon (needs a rc file) and the macOS application bundle image. At the moment I cannot compile on Windows and I don't have access to a Mac at all.

**Comment by Manuel-K** [comment]
Should I use a larger png file?

**Comment by Manuel-K** [comment]
Current progress is in https://github.com/ja2-stracciatella/ja2-stracciatella/tree/minor-launcher-improvements

**Comment by lynxlynxlynx** [comment]
re padding: yes.
The bold is too strong to me, it looked better before. I suggest you instead add more spacing from the upper two elements.

I guess it would also be good to somehow expose game.json editing.

**Comment by Manuel-K** [comment]
I'm not sure about the bold text, too. Most other applications use it for group headers.

**Comment by Manuel-K** [comment]
Not bold and a bit more space:
![launcher_4](https://user-images.githubusercontent.com/775439/32460622-ab807286-c333-11e7-97a5-d5f79d34054b.png)

**Comment by Manuel-K** [comment]
All in all it's probably a better idea to embed a bmp file. That should be much smaller than statically linking libpng and libz.

**Comment by Manuel-K** [comment]
On the other hand we would lose the alpha channel.

**Comment by Manuel-K** [comment]
Alternative: Embed xpm, create `Fl_Pixmap` from embedded xpm, create `Fl_RGB_Image` from `Fl_Pixmap` and use that as icon.

**Comment by Manuel-K** [comment]
Funny: The FLTK version used by Travis lacks the necessary constructor.

**Comment by Manuel-K** [comment]
We use trusty on Travis which has FLTK 1.3.2 and the constructor was added in 1.3.3 released on 03.11.2014.

**Comment by lynxlynxlynx** [comment]
surely there's a ppa with the newer one that we could use.

**Comment by Manuel-K** [comment]
There has to be. But do we seriously want to limit the launcher to FLTK >= 1.3.2 to avoid linking to libpng and libz for the main window icon? I'm not sure that it's a good idea.

**Comment by Manuel-K** [comment]
@selaux @derSteFfi png or xpm?

**Comment by Manuel-K** [comment]
If we add support for selecting mods in the future, it would be nice to include some images as well and they are more likely to be png/jpg than xpm.

**Comment by derSteFfi** [comment]
png sounds fine to me.

**Comment by JAsmine-ja2** [comment]
Time for my cheap mockups :) to show some ideas:

![1](https://user-images.githubusercontent.com/17213795/32494413-de8695a0-c3c1-11e7-8562-9971ce7731a3.png)
![2](https://user-images.githubusercontent.com/17213795/32494414-e0004098-c3c1-11e7-90f2-c6204dd77857.png)

**Comment by Manuel-K** [comment]
Well, not noticing the Ja2 is embarrassing, but at least I didn't write it. ;-)

**Comment by Manuel-K** [comment]
Current version:
![launcher_5](https://user-images.githubusercontent.com/775439/32501231-c762e246-c3d7-11e7-9f41-143ccf88dead.png)

**Comment by Manuel-K** [comment]
How should we set the version number? At compile time would be the easiest solution, but it would be wrong if the user would update only one of the binaries.

Either way, I'm against introducing a separate launcher version number.

**Comment by lynxlynxlynx** [comment]
I wouldn't worry about users managing to update only one. But I also think there's no need to display the version here.

**Comment by Manuel-K** [comment]
Okay. I don't think the version number in the launcher is important.

Then I'll revert the xpm changes and merge the code corresponding with the last screenshot for now.

**Comment by selaux** [comment]
Yes, I also think png will be better. And sorry for the Ja2 😁 .

**Comment by selaux** [comment]
Version number could become interesting for doing a check wether the installed version is also the most recent one. That might be something to include in the launcher as well.

**Comment by selaux** [comment]
I will have a look at the resolution related stuff.

**Comment by Manuel-K** [comment]
It might be a good idea if the launcher had the option to display `ja2.log`.

**Comment by RinneT** [comment]
Just an FYI for the resolution. Custom resolution ratios can cause erroneous rendering.
I realized that when I accidentally set the resolution to something like 1920x1200. It caused some objects to be rendered one tile from the cursor.

**Comment by RinneT** [comment]
Nevermind my last comment. It seems to have been fixed.

**Comment by Manuel-K** [comment]
Let's talk about the proposed `game.json` editor:

Until the discussion in #623 I've assumed that it should be with all bells and whistles i.e. use check boxes and drop down lists to set the values.

Other possibilities would be to open an external text editor or use a simple text editor widget to create an internal minimal text editor.

**Comment by lynxlynxlynx** [comment]
text editor: full power, no need for format change, easier to shoot yourself in the foot, less code
gui abstraction: need for a format change (parser + tooltip/help), harder to screw up, uglier for direct editors

But the main problem is ensuring the file is editable. It would have to be installed in the user's area, so like the main config. Which brings in versioning and migration problems as we add new options. Ugh.

**Comment by Manuel-K** [comment]
Another basic alternative: We could add a short text explaining `game.json` that displays the automatically determined location, e.g. “Open `/path/to/game.json` with a text editor for advanced settings”.

**Comment by lynxlynxlynx** [comment]
KISS, I like it. :+1:

**Comment by JAsmine-ja2** [comment]
A full-bown editor tool for the externalized data would be nice but is probably be a bit over the top.

Imo, it is already a very big step ("the" biggest step) to actually have externalized data. Thus, customizing elements of the game becomes as easy as editing a text file as opposed to hacking the exe.

So, I'm pro @Manuel-K 's idea

**Comment by Manuel-K** [comment]
Getting the location of `game.json` requires some code duplication, a lot of refactoring or linking to SDL2 which is a bad idea. 

I've started the refactoring, but it requires many changes to unrelated code: Pulling out `findObjectCaseInsensitive` and `pathsJoin` from `FileMan` and moving it to a new class `PathTools`. TODO: `FileMan_unittest.cc` has to be split into `FileMan` and `PathTools` parts. `externalizedDataPath` definition in `SGP.cc` has to be accessible too. I'm thinking about adding it to `PathTools` as `getExternalizedDataPath`.

**Comment by selaux** [comment]
Imho we could also move the `PathTools` part to rust. Rust provides path handling stuff in the [stdlib](https://doc.rust-lang.org/std/path/struct.Path.html), so we could get rid of the boost dependency for fileman.

**Comment by Manuel-K** [comment]
Added: Add a tooltip for `data_dir`.

**Comment by lynxlynxlynx** [comment]
For us it's the data dir, but as the report shows, people misunderstand it. I suggest you change the text to "JA2 game directory".

**Comment by Manuel-K** [comment]
Both. Tooltip msg should be something like "this should point to the directory that contains the `data` folder".

**Comment by tanhorn** [comment]
I think would be better "point to game directory"
Actually, I think, that in launcher must be some kind of checkbox for choosing mods?

PS. Just solved another problem with cruses - changed driver for videocard

**Comment by Manuel-K** [comment]
After the update in #693:
![launcher_6](https://user-images.githubusercontent.com/775439/35194039-904fa7f2-feac-11e7-9405-8562abc90c87.png)

**Comment by lynxlynxlynx** [comment]
You mean before, right — the pr changes data to game.

**Comment by Manuel-K** [comment]
Crap. It's the wrong screenshot. Again.

**Comment by Manuel-K** [comment]
Once again:
![launcher_6](https://user-images.githubusercontent.com/775439/35195463-15fc06c2-fec4-11e7-82a1-40ea297a27cc.png)

**Comment by benjaminfoo** [comment]
I don't know if this is called necrobumping, nor a new issue should be created for this - but what are the benefits of using fltk? I think the appearance is a bit outdated - id like to support this project by creating an updated version of the launcher, but I wouldn't use fltk for it.

**Comment by lynxlynxlynx** [comment]
From memory it was chosen because it is crossplatform, stable and doesn't drag too many dependencies with it (eg. compared to pyQt). What did you have in mind?

**Comment by benjaminfoo** [comment]
Well, I'd like to use the launcher the theme / schema of the system, if any provided - this could be realized changing the FLTK scheme (using `Fl::scheme("gtk+");` ) for example. 

**Without applied color-scheme**
![without_scheme](https://user-images.githubusercontent.com/2045944/56231479-deb41380-607e-11e9-869e-5441f1d29888.png)

**With applied color-scheme** 
![with_scheme](https://user-images.githubusercontent.com/2045944/56231488-e1af0400-607e-11e9-9e36-02103b727106.png)

**Comment by lynxlynxlynx** [comment]
Does it gracefully degrade if gtk (1??) isn't present?
But you can also make a case for a different toolkit. It'd be extra interesting if there's something based on sdl, since we already use it and our gui needs are minute.

**Comment by flaviojs** [comment]
Resizing in FLTK is pretty simple: https://www.fltk.org/articles.php?L415

The question is how do you want it to resize.
Imagine the launcher is maximized, would it:
1) expand the space around? the contents would be centered and with the same size
2) expand the space between the sections? (top, bottom-left, bottom-right)
3) scale every widget? would look ugly because text size stays the same (not part of layout)
4) resize in some other way?...

**Comment by lynxlynxlynx** [comment]
I don't think we need to make it resizable in the first place. It does need to be a bit bigger though, to fit the whole label for the scaling options.

**Comment by flaviojs** [comment]
Ok, I'll ignore resizing.

Then what would you prefer for a game.json editor? (which I just noticed is invalid json because of the comments)
1) add tabs and put the editor in a new tab?
2) open a new window?
3) open the file with the default app or open the file location? (probably has portability issues)

**Comment by lynxlynxlynx** [comment]
See the previous discussion (hidden by default) — the consensus was to just display the path and let the power users deal with the editing themselves.

We specifically use a json implementation that handles comments, so it's more user friendly.

**Comment by flaviojs** [comment]
Totally missed that, the first message needs updating to avoid confusion. (or this issue needs retiring)

Will get on that after the guess version code is merged.
I'm thinking of a non editable input field below the "Game Version" input.

While I'm at it I'll also make sure the scaling mode text is fully visible.

**Comment by flaviojs** [comment]
I've been playing around with fluid and ended up with this layout, what do you think of this mockup?
![with_game_settings](https://user-images.githubusercontent.com/1009600/57996824-96df3c80-7ac1-11e9-9121-763c89385ed9.png)
You should note that I'm also combining the resolution inputs.

**Comment by selaux** [comment]
While this looks cleaner than now, it still contains a lot of inputs in a small space, without any space for explanation. I personally would love to have a layout like the [openmw launcher](https://openmw.org/), with multiple tabs, so we can add some more options without having to change design.

![https://user-images.githubusercontent.com/6200170/35781519-c3c6c3ae-09b0-11e8-9ece-636d185ebf59.png](https://user-images.githubusercontent.com/6200170/35781519-c3c6c3ae-09b0-11e8-9ece-636d185ebf59.png)

**Comment by flaviojs** [comment]
Ok, so something like this:
![tab_play](https://user-images.githubusercontent.com/1009600/58014219-3c110980-7af0-11e9-8185-17ede60211d6.png)
![tab_data](https://user-images.githubusercontent.com/1009600/58014221-403d2700-7af0-11e9-995b-bdaaf2af3e12.png)
![tab_settings](https://user-images.githubusercontent.com/1009600/58014223-429f8100-7af0-11e9-86c5-3e1be0473af2.png)
If it had something for mods, it would go in the Data tab.
